### PR TITLE
Chart.js and Google Chart prototypes

### PR DESCRIPTION
**Pull request by EmmaLRussell**
In this branch I've added into the `bar` directory a page with prototypes showing Chart.js and Google chart implementations, with a dropdown for Indicator which will update the view. These both share the sample data from `/processedData/data.ts`, which is in something like the form that comes back from hintr, and also metadata which is something like what might come back from hintr's configuration (although that will give more options to the user).

For both chart types, there is also a component with hardcoded data in the `hardcodedData` folder, which shows what its data format 'naturally' looks like. (These aren't actually present in index.html.) 

For both `processedData` components, the code transforming the sample data into the chart's required format is in the `processedOutputData` computed property. 

Let me know which you prefer! My own feelings:

**Chart.js win**
- Looks nicer - deals more elegantly with limited space
- Already working for us in datavis tool with no obvious performance issues

**Google win**
- Updates data automatically, unlike Chart js, which requires renderChart to be called. 
- Supports error bars (intervals) natively - chart.js requires a plugin - and shows interval values automatically on mouseover. 

**Call it a draw**
- data format. Both Google's tabular format and Chart.js separate labels+datasets are fairly awkward. I ended up putting Google's rows into an map to find them easily - could do the same with Chart.js's datasets to avoid the possibly quite expensive filter. I think I actually slightly prefer chart.js for being more structured.

### CODE REVIEW

File: bar/processedData/data.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    {age_group: '30-34', sex: 'male', indicator: '3', mean: 0.15, high: 0.17, low: 0.13},
+    {age_group: '35-39', sex: 'male', indicator: '3', mean: 0.19, high: 0.23, low: 0.12},
+    {age_group: '40-44', sex: 'male', indicator: '3', mean: 0.22, high: 0.24, low: 0.2},
+    {age_group: '45-49', sex: 'male', indicator: '3', mean: 0.28, high: 0.29, low: 0.25},
+    {age_group: '50-54', sex: 'male', indicator: '3', mean: 0.25, high: 0.28, low: 0.13},
+    {age_group: '55-59', sex: 'male', indicator: '3', mean: 0.11, high: 0.15, low: 0.09}
+];
+
+//TO COME FROM METADATA / USER SELECTIONS
+export const chartMetadata = {
```
Comment (hillalex) [inline bar/processedData/data.ts:57]: Can we split this up into the bit that's returned from the API and the bit that's selected by the user, to come up with a spec for the API? Then the other bit is metadata about what the filters themselves are...

**Review by hillalex — COMMENTED** [review]
I prefer chartjs both visually and codewise, and it seems much quicker to render also - a clear winner! Want to take that one forward and flesh out the prototype to derive the full API spec?

File: bar/processedData/data.ts (hunk 1 above)
Comment (EmmaLRussell) [reply bar/processedData/data.ts:57]: > I prefer chartjs both visually and codewise, and it seems much quicker to render also - a clear winner! Want to take that one forward and flesh out the prototype to derive the full API spec?

Oh good, me too!

Comment (EmmaLRussell) [reply bar/processedData/data.ts:57]: Yep, going to put together some sample metadata and types next, and plug it into this.

**Comment by EmmaLRussell** [comment]
I've updated this to a somewhat simplified, but functional approximation of what we need from the bar chart in terms of filtering and setting axes, with metadata and related types defined.

**Comment by r-ash** [comment]
This looks great - I might try and show it to Jeff and get some feedback from him too.

I think you mentioned that the behaviour when selecting same disaggregate by & x-axis is a bit funny with bar widths. 

I like labelling the filters as x-axis/disaggregate by, only one label is applied when the same filter is chosen for both of those.

From speaking to Jeff I think the filters for region we expect to look similar to that in the maps, for the age filters he wants to be able to specify arbitrary groups of those (some of which may be size 1). 

Something like

```
15 - 49
15+
15 - 64
5 year age groups
  0 - 4
  5 - 9
  10 - 14
  15 - 20
  ...
Coarse age groups
  0 - 14
  15 - 24
  25 - 34
  ...
```

Where is users choose a top level then it selects all below them (or they can select at the lower level). For the groups of 1 level - these have values associated with them, the other aggregations don't. We could model that as a group with a single value and have the UI not show the sub levels when there is only 1 or something.

**Review by hillalex — COMMENTED** [review]
This looks really great. I've put my proposed suggestions for breaking up the code and making more extensive use of computed values in a separate directory: https://github.com/mrc-ide/vue-prototypes/pull/2/commits/0708d6ca6576e45a2506821e126526b51634b76f

In terms of the schema, my only suggestion is that it feels more natural to me that the filters and their options would come in [one object](https://github.com/mrc-ide/vue-prototypes/pull/2/commits/0708d6ca6576e45a2506821e126526b51634b76f#diff-0fc39a7c6ac8a823a5c794117fb934a3R16) rather than having the options contained in a [separate object](https://github.com/mrc-ide/vue-prototypes/pull/2/files#diff-7eb8e595fe21686b3b7b3fda53924ae6R32) but perhaps I'm missing something? Is there a reason to return these separately? In any case, if there is, they could be combined in the front-end before passing them to the component.

Re the look and feel, I think it's great. The only change I would make is to the "selecting nothing in the multiselects = showing everything on the graph" behaviour. It feels confusing to me, I understand that it's come from thinking of and calling the select widgets "filters", but that's really just a matter of terminology, and from a UI point of view I think it's more natural to think of the data being shown as directly corresponding to the items selected. But of course this is just my opinion, really would be good to show some naive users and see which feels more natural to them.

File: bar/useMetadata/data.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    {region: '3', age_group: '25-29', sex: 'male', indicator: '3', mean: 0.28, high: 0.35, low: 0.27},
+    {region: '3', age_group: '30-34', sex: 'male', indicator: '3', mean: 0.24, high: 0.27, low: 0.23},
+    {region: '3', age_group: '35-39', sex: 'male', indicator: '3', mean: 0.23, high: 0.33, low: 0.22},
+    {region: '3', age_group: '40-44', sex: 'male', indicator: '3', mean: 0.31, high: 0.34, low: 0.3},
+    {region: '3', age_group: '45-49', sex: 'male', indicator: '3', mean: 0.16, high: 0.22, low: 0.15},
+    {region: '3', age_group: '50-54', sex: 'male', indicator: '3', mean: 0.14, high: 0.18, low: 0.13},
+    {region: '3', age_group: '55-59', sex: 'male', indicator: '3', mean: 0.24, high: 0.25, low: 0.19},
+];
+
+export const colors = [
```
Comment (hillalex) [inline bar/useMetadata/data.ts:234]: Is the idea that these come from the API also?

Comment (EmmaLRussell) [reply bar/useMetadata/data.ts:234]: Rob negotiated this one, but as I understand it, to keep things simple, Jeff has a agreed to a fixed colour sequence to use, so we can just keep it in our code unless he changes his mind!

**Comment by EmmaLRussell** [comment]
So the reason the filter options and definitions are in separate objects is because they come from different responses. The filter options are part of the data response objects (ModelRunResponse in this case) - these include the *actual* values of region, age group etc for a given dataset. These are shared for that response across all plot types. 

The filter definitions are part of the metadata for a particular type of plot and are meant to be fixed between all responses (for a country) and define the filters to include, the display label and anything else related to that plot type.

I agree means the code using these structures is quite fragmented. Maybe we could create a getter which combines these data structures from the API into a single one which is easier to wrangle...

**Comment by r-ash** [comment]
> So the reason the filter options are in a separate object is because they are part of the response objects (ModelRunResponse in this case) - these include the _actual_ values of region, age group etc for a given dataset. These are shared for that response across all plot types.
> 
> The filter definitions are part of the metadata for a particular type of plot and are meant to be fixed between all responses (for a country) and define the filters to include, the display label and anything else related to that plotting.
> 
> I agree means the code using these structures is quite fragmented. Maybe we could create a getter which combines these data structures from the API into a single one which is easier to wrangle...

We could remove the `plottingMetadata` endpoint entirely and always return that with the request which gets the data and filters.

**Comment by EmmaLRussell** [comment]
> > So the reason the filter options are in a separate object is because they are part of the response objects (ModelRunResponse in this case) - these include the _actual_ values of region, age group etc for a given dataset. These are shared for that response across all plot types.
> > The filter definitions are part of the metadata for a particular type of plot and are meant to be fixed between all responses (for a country) and define the filters to include, the display label and anything else related to that plotting.
> > I agree means the code using these structures is quite fragmented. Maybe we could create a getter which combines these data structures from the API into a single one which is easier to wrangle...
> 
> We could remove the `plottingMetadata` endpoint entirely and always return that with the request which gets the data and filters.

Don't mind. The getter approach would work too. Maybe we could start with the getter approach to develop a data structure which is good to work with out of the responses we have, then later we could move to getting that data structure direct from hintr.

**Comment by EmmaLRussell** [comment]
> Re the look and feel, I think it's great. The only change I would make is to the "selecting nothing in the multiselects = showing everything on the graph" behaviour. It feels confusing to me, I understand that it's come from thinking of and calling the select widgets "filters", but that's really just a matter of terminology, and from a UI point of view I think it's more natural to think of the data being shown as directly corresponding to the items selected. But of course this is just my opinion, really would be good to show some naive users and see which feels more natural to them.

I see what you mean but I tend to think it's actually more usable as it is, since it's probably the default to want to see all values initially, and it would keep the UI cleaner if that default didn't mean pre-selecting everything. Would it be clearer if the placeholder text on the empty multi-select said 'Show all' or similar?

**Comment by hillalex** [comment]
> > So the reason the filter options are in a separate object is because they are part of the response objects (ModelRunResponse in this case) - these include the _actual_ values of region, age group etc for a given dataset. These are shared for that response across all plot types.
> > The filter definitions are part of the metadata for a particular type of plot and are meant to be fixed between all responses (for a country) and define the filters to include, the display label and anything else related to that plotting.
> > I agree means the code using these structures is quite fragmented. Maybe we could create a getter which combines these data structures from the API into a single one which is easier to wrangle...
> 
> We could remove the `plottingMetadata` endpoint entirely and always return that with the request which gets the data and filters.

I think this makes sense. The filter definitions and filter options are always used together, so from the front-end's point of view there is no point in having 2 endpoints - they are always combined at the point of use.

**Comment by hillalex** [comment]
> > Re the look and feel, I think it's great. The only change I would make is to the "selecting nothing in the multiselects = showing everything on the graph" behaviour. It feels confusing to me, I understand that it's come from thinking of and calling the select widgets "filters", but that's really just a matter of terminology, and from a UI point of view I think it's more natural to think of the data being shown as directly corresponding to the items selected. But of course this is just my opinion, really would be good to show some naive users and see which feels more natural to them.
> 
> I see what you mean but I tend to think it's actually more usable as it is, since it's probably the default to want to see all values initially, and it would keep the UI cleaner if that default didn't mean pre-selecting everything. Would it be clearer if the placeholder text on the empty multi-select said 'Show all' or similar?

Maybe! Could try changing the placeholder to that, yeah.

**Pull request merged**